Print row types when they are a type constructor

smuenzel opened the pull request:
This improves error messages, by not hiding types that are necessary to interpret the error:
```
Error: This expression has type "< bar : int; foo : int; .. >"
       but an expression was expected of type "'a"
       The type constructor "$1" would escape its scope
```

becomes

```
Error: This expression has type "< bar : int; foo : int; .. as $1>"
       but an expression was expected of type "'a"
       The type constructor "$1" would escape its scope
```

smuenzel committed:
Print row types when they are GADT constructors

smuenzel commented on testsuite/tests/typing-gadts/test.ml:
```diff
@@ -1081,9 +1081,10 @@ let g (type t) (x:t) (e : t int_foo) (e' : t int_bar) =
        ^
-Error: The value "x" has type "t" = "< foo : int; .. >"
+Error: The value "x" has type "t" = "< foo : int; .. > as $0"
        but an expression was expected of type "< bar : int; foo : int >"
-       Type "$0" = "< bar : int; .. >" is not compatible with type "< bar : int >"
+       Type "$0" = "< bar : int; .. > as $1" is not compatible with type
```
hmmmm there might be something wrong with variable naming here

smuenzel replied:
I now think this is actually correct, `$0` and `$0` are the same, in the first case, we just see it in the `as` clause (but we don't expand it) and in the second case it is expanded. 
Since we're failing in the middle of unification, we haven't yet propagated that it could be `< foo : int; bar : int; ..>`.
The abstract row check is happening first.

Octachron commented on typing/out_type.ml:
```diff
@@ -1176,7 +1176,26 @@ and tree_of_typobject mode fi nm =
+      | Some ty ->
+          match get_desc ty with
+          | Tconstr (Pident p, _, _)
+            when String.ends_with ~suffix:"#row" (Ident.name p) ->
```
This is `Btype.is_row_name` (with a better implementation).

hhugo replied:
`Btype.is_row_name` could use `String.ends_with` instead of `String.sub` and equality

Octachron commented on testsuite/tests/typing-gadts/test.ml:
```diff
@@ -1067,9 +1067,9 @@ type _ int_bar = IB_constr : < bar : int; .. > int_bar
+Error: The value "x" has type "t" = "< foo : int; .. > as $0"
        but an expression was expected of type "< foo : int >"
-       Type "$0" = "< bar : int; .. >" is not compatible with type "<  >"
+       Type "$0" = "< bar : int; .. > as $1" is not compatible with type "<  >"
```
I am not fond of this presentation because it might look like the `as` is giving a name to the object type rather than the row variable while the intended meaning is closer to
```
Type "$0" = "< bar : int; (.. as $1) >" is not compatible with type "<  >"
```

smuenzel replied:
Yeah, that makes sense. Should we attempt to change our existing printing as well? 
```
type t = < bar : int; ..>;;
Error: A type variable is unbound in this type declaration.
       In type < bar : int; .. > as 'a the variable 'a is unbound
```

smuenzel replied:
Very interesting!

My confusion in the test in `typing-gadts/test.ml` is actually coming from just the issue you mentioned. In that case `$0` is a row variable, and `$1` is the continuation of that row variable. When we print `$0`, we actually know that it's a row variable (because it's a `Tfield` and not a `Tobject` -- `type_desc` actually contains two different kinds of types).

So if we extend the syntax of types in error messages to distinguish between objects and "rows", we would get something like:
```
Error: The value "x" has type "t" = "< foo : int; .. as $0>"
       but an expression was expected of type "< foo : int >"
       The row variable "$0" = "< ..; bar : int; .. as $1>"
       is not compatible with the row variable "< ..;  >"
```

smuenzel replied:
I started looking at doing this, however, I'm having some trouble with the type proxies (specifically, distinguishing between the type and the row variable, while still printing everything correctly.
I came up with the following to add in a comment somewhere, is it correct?
```ocaml
(* We use proxies to find shared parts of type expressions/name variables.
   A proxy of a type expression represents the type expression itself.
   However, the difficulty is with open variants and objects -- in that case, we
   don't have a guarantee that two types that are the same will actually compare
   as equal when iterating through the type expression during printing.
   In that case, we use the row variable as a proxy for the whole type.
   This, however, means, we can no longer distinguish between the row variable
   and the type containing the row variable.
*)
```

I guess the current PR can be achieved without looking at this, but I want to clear up my confusion about how this is working.

garrigue replied:
This looks correct: while two types sharing the same row variable are not necessarily physically equal, they are semantically equal, so one does not want to distinguish in printing.
(To be fully correct, there are some subtle cases for principality, where we still want the structure to be unshared to allow seeing a type as "known", but this is orthogonal to the rest of the type checking).
As for the question of how to print the row variable itself, it was considered a non-issue, as row variables are internal artefacts that cannot be directly mentioned. In practice the distinction should come from the context. I agree however that GADTs, by allowing locally abstract types to be instantiated both by the whole type and the row variable, make this issue more acute. (But locally abstract row variables are not officially supported and do not work correctly in many cases.)

garrigue replied:
Also, concerning printing, I am not opposed to writing `.. as $1` when the row variable is a locally abstract type, since anyway the syntax `<sthg> as $1` is not supported by the parser.
However, I do not see the point of the front `..` in `< ..; bar : int; .. as $1 >`. A row can only be a number of fields, followed by a row variable.

garrigue commented:
This looks like a nice improvement.
Note that in ocaml row variables are not given first class status.
This is the reason why `as` binds on the whole type rather than the internal row variable.
This is usually ok because we have the invariant that a row variable cannot be shared between different types, so the alias ends up representing both the row variable and the whole type.
There are indeed a few cases where this is confusing, but changing this requires thinking about design issues.

smuenzel commented:
Thanks for the review! I addressed your comments.

Regarding the `< ..` notation, I wanted to distinguish cases where we have a complete object (wrapped in `Tobject`) vs a list of fields. But looking at the error messages again, I agree that this isn't particularly useful.

Octachron commented on typing/outcometree.mli:
```diff
@@ -67,12 +67,12 @@ type out_type_param = {
   | Otyp_abstract
   | Otyp_open
-  | Otyp_alias of {non_gen:bool; aliased:out_type; alias:string}
+  | Otyp_alias of {aliased:out_type; alias:out_alias}
```
The updated representation for aliases is currently not produced in the PR, isn't it?

smuenzel replied:
Yes, removed!

Octachron commented on testsuite/tests/typing-modules/inclusion_errors.ml:
```diff
@@ -1298,7 +1298,7 @@ Error: Signature mismatch:
          type t = < b : int >
        is not included in
-         type t = private < a : int; .. >
+         type t = private < a : int; .. as t >
```
This error message is leaking an implementation detail of the inclusion test: we test the inclusion
`type t = < x:int; y:int >` ≤ `type t = < z:float; .. as t#row >` in a context with the additional equations 
`t_rhs=t_lhs` and `t_rhs#row` = `t_lhs`. 

I would propose to add an additional preprocessing step in `prepare_decl` to detect this situation where the manifest of a type declaration contains an object row variable which matches the type constructor being declared.

smuenzel replied:
Thanks! done

Octachron commented on typing/out_type.ml:
```diff
@@ -1310,14 +1330,26 @@ let prepare_decl id decl =
+                        newgenty (Tfield (name, kind, tyf, update tyrest))
+                    | _ -> newgenvar ()
+                  in
+                  newgenty (Tobject (update fields, cl))
```
As far I can see, it is fine in this case to `set_type_desc` the object row to `Tvar None`.

Octachron commented on typing/ctype.mli:
```diff
@@ -127,6 +127,7 @@ val associate_fields:
         (string * field_kind * type_expr) list *
         (string * field_kind * type_expr) list
 val opened_object: type_expr -> bool
+val object_row: type_expr -> type_expr
```
I missed the fact that this function was not exposed, sorry. There is already `Btype.row_of_type` which achieve the same result for `Tobject _`.

smuenzel replied:
Fixed.

PS: Should we make a PR that makes sure we only have one version of these similar functions?

Octachron replied:
I tend to agree that removing redundant functions would be nice.

smuenzel committed:
Better Btype.is_row_name

Octachron commented on typing/out_type.ml:
```diff
@@ -1209,6 +1224,11 @@ and tree_of_package mode {pack_path; pack_constraints} =
+and tree_of_type_path p =
+  let (p', s) = best_type_path p in
+  let p'' = if (s = Id) then p' else p in
+  tree_of_best_type_path p p''
```
Moving this function here is strange and potentially confusing for future reader since the function is not used in the recursive group, nor does it have any dependency within the group. I would propose to either let the function in its current position or move it up just after the definition of `tree_of_best_type_path`.

smuenzel replied:
Move up, thanks

Octachron commented on typing/btype.ml:
```diff
@@ -283,7 +283,7 @@ let is_row_name s =
   (* PR#10661: when l=4 and s is "#row", this is not a row name
      but the valid #-type name of a class named "row". *)
-  l > 4 && String.sub s (l-4) 4 = "#row"
+  l > 4 && String.ends_with ~suffix:"#row" s
```
This is a good refactoring even if a bit unrelated, let's keep it.

Octachron reviewed (approved):
The current state looks good and useful. I have one minor nitpicking comment left, I will let you rewrite the history to your preference since there is a quite a lot of back-and-forth currently.

smuenzel commented:
Thanks for the review, I cleaned up the history now. Let's see if the testsuite agrees.
